From 980d7494dcdaa380740e6280b42ebc461f58e9b3 Mon Sep 17 00:00:00 2001
From: Nicolas Le Bayon <nicolas.le.bayon@foss.st.com>
Date: Mon, 3 Feb 2025 14:07:17 +0100
Subject: [PATCH 783/954] fix(st-crypto): increase RNG robustness for STM32MP21

On STM32MP21, enabling RNG starts the hardware execution of the BSEC
HWBUS which will generate the DDRcrypt encryption key (RISAF4 for DDR).
It is crucial that no seed error occurs during this phase.
Reducing the RNG maximum noise clock frequency increase the RNG
robustness without impacting the general boot time.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 9aa6840ca989136d16a779cda6fd3d024d9e1d3b fix(st-crypto): increase RNG robustness for STM32MP21]
Signed-off-by: Nicolas Le Bayon <nicolas.le.bayon@foss.st.com>
Change-Id: I2ae49be65bc20fdb234dccd7076411be2414673f
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/433842
Reviewed-by: Nicolas LE BAYON <nicolas.le.bayon@st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
Tested-by: Nicolas LE BAYON <nicolas.le.bayon@st.com>
---
 drivers/st/crypto/stm32_rng.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/drivers/st/crypto/stm32_rng.c b/drivers/st/crypto/stm32_rng.c
index dd7bcfcbdf..711839ffa1 100644
--- a/drivers/st/crypto/stm32_rng.c
+++ b/drivers/st/crypto/stm32_rng.c
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2022-2024, STMicroelectronics - All Rights Reserved
+ * Copyright (c) 2022-2025, STMicroelectronics - All Rights Reserved
  *
  * SPDX-License-Identifier: BSD-3-Clause
  */
@@ -68,6 +68,7 @@
 #define RNG_NIST_CONFIG3		0xDU
 #define RNG_HTCFG_CONFIG		0x0000969DU
 #define RNG_NSCFG_CONFIG		0x0002B5BBU
+#define RNG_MAX_NOISE_CLK_FREQ		48000000U
 #elif STM32_RNG_VER_MINOR == 3
 /* MP25 and MP23 default values */
 #define RNG_NIST_CONFIG1		0x8FU
@@ -75,6 +76,7 @@
 #define RNG_NIST_CONFIG3		0xEU
 #define RNG_HTCFG_CONFIG		0x00006688U
 #define RNG_NSCFG_CONFIG		0x0002E649U
+#define RNG_MAX_NOISE_CLK_FREQ		48000000U
 #elif STM32_RNG_VER_MINOR == 4
 /* MP21 default values */
 #define RNG_NIST_CONFIG1		0xFU
@@ -82,13 +84,12 @@
 #define RNG_NIST_CONFIG3		0xFU
 #define RNG_HTCFG_CONFIG		0x0000AAC7U
 #define RNG_NSCFG_CONFIG		0x000001FFU
+#define RNG_MAX_NOISE_CLK_FREQ		4000000U
 #else
 #error "Please define STM32_RNG_VER_MINOR"
 #endif
 #endif
 
-#define RNG_MAX_NOISE_CLK_FREQ		48000000U
-
 struct stm32_rng_instance {
 	uintptr_t base;
 	unsigned long clock;
-- 
2.53.0

